### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by lukacsja**
[DEMO LINK](https://lukacsja.github.io/react_todo-app-with-api/)

Everything looks to work as expected except that on doubleclick I need to click one more time to be able to edit the field. 
I'm open to any suggestions for that and everything else.

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,273 @@
...
+            setFilteredTodos(todosFromServer);
+            setLoadingTodoIds([]);
+          })
+          .catch(() => {
+            setError(Errors.CantUpdate);
+          });
+      });
+  };
+
+  const toggleAllTodoStatus = () => {
```
Comment (naziamoff) [inline src/App.tsx:78]: very huge method, try to divide it into the smaller ones

File: src/components/Footer.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  completedTodos,
+  filterByAll,
+  filterByActive,
+  filterByCompleted,
+  selectedFilter,
+  clearCompletedTodos,
+}) => {
+  const hasFilterAll = selectedFilter === Filters.All;
+  const hasFilterActive = selectedFilter === Filters.Active;
+  const hasFilterCompleted = selectedFilter === Filters.Completed;
```
Comment (naziamoff) [inline src/components/Footer.tsx:26]: redundant variables, you can do it right in the classnames

File: src/components/Main.tsx [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                  type="text"
+                  className="todo__title-field"
+                  placeholder="Empty todo will be deleted"
+                  value={newTitle}
+                  onChange={changeEditField}
+                  onBlur={(event) => {
+                    event.preventDefault();
+                    handleEditSubmit(todo.id, todo.title);
+                    saveEditedTitle(todo.id, newTitle, setEditingTodoId);
+                  }}
```
Comment (naziamoff) [inline src/components/Main.tsx:84]: worth a handler

File: src/components/Main.tsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+          />
+        </label>
+
+        <div className="modal overlay">
+          <div className="modal-background has-background-white-ter" />
+          <div className="loader" />
+        </div>
+      </div> */}
+
+      {/* This todo is in loadind state */}
```
Comment (naziamoff) [inline src/components/Main.tsx:136]: do you need those comments? Seems it is the last part of the task

**Review by naziamoff — CHANGES_REQUESTED** [review]
great job, Luka 🔥 
Let's polish and you're good

**Comment by lukacsja** [comment]
> about your question - there's a method for focusing on the element. Try to google, and if you won't succeed in it - let me know and I'll help

thank you. finally managed to implement it. also it doesnt save when the title doesnt change now.

File: src/App.tsx (hunk 1 above)
Comment (lukacsja) [reply src/App.tsx]: i broke it into 4 functions. not sure if its more readable tho

File: src/App.tsx [hunk 5]
```diff
@@ -1,24 +1,311 @@
...
+      });
+  };
+
+  const getUpdatedTodos = (
+    todosToUpdate: Todo[],
+    everyTodoCompleted: boolean,
+    noTodoCompleted: boolean,
+    someTodosCompleted: boolean,
+  ) => {
+    const updatedTodos: Todo[] = [];
```
Comment (meforoma) [inline src/App.tsx:84]: you can provide an initial value in args area
```suggestion
    todosToUpdate: Todo[] = [],
    everyTodoCompleted: boolean,
    noTodoCompleted: boolean,
    someTodosCompleted: boolean,
  ) => {
```

File: src/App.tsx [hunk 6]
```diff
@@ -1,24 +1,311 @@
...
+    todosToUpdate.forEach(todo => {
+      let updatedTodo = { ...todo };
+
+      if (someTodosCompleted && !todo.completed) {
+        updatedTodo = { ...todo, completed: !todo.completed };
+      }
+
+      if (everyTodoCompleted || noTodoCompleted) {
+        updatedTodo = { ...todo, completed: !todo.completed };
+      }
```
Comment (meforoma) [inline src/App.tsx:95]: could be shorter - update props when constructing updatedTodo
```suggestion
      const completed = (someTodosCompleted && !todo.completed)
        || (everyTodoCompleted || noTodoCompleted)
      const updatedTodo = {
        ...todo,
        completed,
      };
```

File: src/App.tsx [hunk 7]
```diff
@@ -1,24 +1,311 @@
...
+    const loadingTodoIdsToAdd: number[] = [];
+
+    todosToLoad.forEach(todo => {
+      if (someTodosCompleted && !todo.completed) {
+        loadingTodoIdsToAdd.push(todo.id);
+      }
+
+      if (everyTodoCompleted || noTodoCompleted) {
+        loadingTodoIdsToAdd.push(todo.id);
+      }
```
Comment (meforoma) [inline src/App.tsx:118]: you can combine evaluations like in prev comment
```suggestion
      if ((someTodosCompleted && !todo.completed)
        || (everyTodoCompleted || noTodoCompleted)) {
        loadingTodoIdsToAdd.push(todo.id);
      }
```

File: src/components/Main.tsx [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                  onChange={changeEditField}
+                  onBlur={(event) => {
+                    event.preventDefault();
+                    handleEditSubmit(todo.id, todo.title);
+                  }}
+                  onKeyDown={(event) => {
+                    if (event.key === 'Escape') {
+                      setEditingTodoId(0);
+                    }
+                  }}
```
Comment (meforoma) [inline src/components/Main.tsx:95]: let's keep it as short as possible
```suggestion
                  onBlur={handleBlur}
                  onKeyDown={handleEscape}
```

**Review by meforoma — CHANGES_REQUESTED** [review]
let's address few comments and move on

File: src/App.tsx (hunk 6 above)
Comment (lukacsja) [reply src/App.tsx:95]: thank you for your comments, i created a nice and common handlerFunction for both 'blur' and ' form submit' but with this one i had hard times shortening it.
the above described case updates all of them even when there are some completed and some active todos so i think i cant go for it. 
or i am just missing something

Comment (meforoma) [reply src/App.tsx:95]: oh, so you mean that `everyTodoCompleted || noTodoCompleted` are dependent from `someTodosCompleted && !todo.completed`?
if so, you could carry on.

I think you should just try it and test out edge-cases that should be handled by the logic.

Have you tried to apply the suggestion? And test it on the rendered page?

Comment (lukacsja) [reply src/App.tsx:95]: yes of course but then i realized that i probably need either switch or if statements because sometimes on only want to update some of them (both on the page and in the api)

**Review by meforoma — APPROVED** [review]
nice 👍🏼
